### ISSUE TITLE
Standard way of sharing architectures?

### ISSUE THREAD

**Issue by SamuelMarks** [tensorflow/tensorflow#30896]
**System information**
- TensorFlow version (you are using):
1.14, and sometimes the latest v2 pre-release
- Are you willing to contribute it (Yes/No):
Yes

**Describe the feature and the current behavior/state.**
Found a number of interesting architectures on GitHub. Starting making PRs to them, to support Python 2.7–3.6, and soon for proper setuptools integration.

The [Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) API, and [tfds in general](https://www.tensorflow.org/datasets/api_docs/python/tfds) are both useful approaches for preparing data, and exposing it for others to consume.

However, AFAIK, there is no standard way of sharing, say: Keras models. Internally have moved to [`gin`](https://github.com/google/gin-config), but not sure if that's the best approach for all.

**Will this change the current api? How?**
Introduce a new module, similar to how tfds was introduced for preparing and packaging datasets.

**Who will benefit with this feature?**
Everyone. Especially those that want to combine multiple architectures from different sources—or in my scenario—to experiment with different architectures.

**Any Other info.**
Concept:
```python
import some_broke_arch
import other_neat_arch
import horrible_v_arch

model   = some_broke_arch.get_arch(   **standard_arch_params  )
metrics = other_neat_arch.get_metrics(**standard_metric_params)
loss    = horrible_v_arch.get_loss(   **standard_loss_params  )

model.compile(loss=loss, optimizer=keras.optimizers.RMSprop, metrics=metrics)
print(model.summary())
# &etc.
````

**Comment by rmothukuru** [comment]
@SamuelMarks,
Sorry for the delayed response. Can you please let us know if the **`Pre-Trained Models`** mentioned in [tf.keras.applications](https://www.tensorflow.org/api_docs/python/tf/keras/applications) is what you are looking for? Thanks!

**Comment by SamuelMarks** [comment]
@rmothukuru No, what I suggested two years ago is a general interface/concept that is to be implemented by each of architecture, metric, loss, &etc.; and also for the model. I am aware of `tf.keras.applications`; having used it for years (back when there were multiple backends for Keras). `include_top`, `weights`, `pooling`, `classes`, `classifier_activation` are common configuration parameters.

When you read a research paper / blog post / Jupyter Notebook they include the exact combination of parameters—and other building-blocks—required to replicate their results. However, if one wants to build an ensemble of 10 of the best networks for solving <problem>, one would have to manually glue them all together.

If one wanted to contribute to every academic open-source repository, include setuptools, dependencies, and expose these as proper Python modules (rather than the common Jupyter format), such that a discoverable set of packages can be exposed (e.g., under a custom classifier on [PyPI](https://pypi.org)… or on some new zoo.tensorflow.org type domain), and that these could then be easily swapped with your provided transfer learning models (out of`tf.keras.application`), and combined into ensembles, and all-in-all compared against each other to improve ones own experiments.

Enough intricacies need to be exposed so you can replace the optimizer|loss|hyperparameter|etc. of a known model to see if your new configuration gives a better result on your dataset.

I don't want to define my own format, as then I'll be fighting others for the best API. I would rather TensorFlow put up a guide and some official interfaces/types that must be implemented by each model in the zoo, such that I can go on my way with these large number of contributions to each open-source—and potentially useful—TensorFlow model.

What do you think?

**Comment by rmothukuru** [comment]
@SamuelMarks,
Thank you for the detailed  explanation. Leaving it open.

**Comment by bhack** [comment]
@SamuelMarks I suggest to start a thread in [our forum](https://discuss.tensorflow.org/)

**Comment by yarri-oss** [comment]
@SamuelMarks following up from todays ML Community Day, did you want to author an RFC instead of filing an issue: https://github.com/tensorflow/community/blob/master/governance/TF-RFCs.md

**Comment by theadactyl** [comment]
I second @bhack on creating a forum thread about this proposal to find an RFC sponsor.

**Comment by SamuelMarks** [comment]
Good idea. I spoke with @fchollet a couple of days ago about this… François Chollet: do you want to be my RFC sponsor?

**Comment by jvishnuvardhan** [comment]
@SamuelMarks Are you still interested in contributing? If yes, please feel free to open a PR in  [keras-team/keras repo.](https://github.com/keras-team/keras/issues) repository.

Please note that Keras development moved to keras-team/keras repository to focus on only keras. Thanks!

Please let me know if you are busy, i can create this PR in keras-team/keras (with your content). Thanks

**Comment by SamuelMarks** [comment]
@jvishnuvardhan Yes: I am still interested in contributing. Okay, I'll cross-post there [and to the [RFC repo](https://github.com/tensorflow/community/tree/master/rfcs)?].

**Comment by bhack** [comment]
I suggest you to post this idea to the forum before opening a new RFC PR as you need to try to collect interested sponsors for the PR

**Comment by bhack** [comment]
With the forum I  mean https://discuss.tensorflow.org

**Comment by SamuelMarks** [comment]
@bhack Yep I had posted on that keras-team org before I saw your reply to post on the forum. The forum is moderated so the topic only just got through the gates:
https://discuss.tensorflow.org/t/standard-way-of-sharing-architectures-e-g-to-pypi-org/6390

Upvote if you please

**Comment by kumariko** [comment]
@SamuelMarks If you have already posted this issue in [TensorFlow Forum](https://discuss.tensorflow.org/t/standard-way-of-sharing-architectures-e-g-to-pypi-org/6390/3) please move this to close status ? Thanks

**Comment by jvishnuvardhan** [comment]
@SamuelMarks We will track the progress in keras-team/keras repo. I am closing this issue here.Thanks
